ccvm_simulators/solvers/pumped_langevin_solver.py

```diff
@@ -0,0 +1,637 @@
+        evolution_step_size,
+        samples_taken,
+    ):
+        """Solves the given problem instance using the original Langevin solver.
+
+        Args:
+            problem_size (int): instance size.
+            batch_size (int): The number of times to solve a problem instance
+            device (str): The device to use for the solver. Can be "cpu" or "cuda".
+            S (float): Saturation bound.
+            pump (float): Pump field strength.
+            dt (float): Simulation time step.
+            iterations (int): number of steps.
+            sigma (float): diffusion constant.
+            pump_rate_flag (bool): Whether or not to scale the pump rate based on the
+            iteration number.
+            feedback_scale (float): feedback scale.
+            evolution_step_size (int): If set, the c/s values will be sampled once
+                per number of iterations equivalent to the value of this variable.
+                At the end of the solve process, the best batch of sampled values
+                will be written to a file that can be specified by setting the evolution_file parameter.
+            samples_taken (int): sample slice.
+
+        Returns:
+            c (tensor): random variable.
+        """
+        # Initialize tensor variables on the device that will be used to perform the
+        # calculations
+        c = torch.zeros((batch_size, problem_size), dtype=torch.float, device=device)
+        wiener_dist_c = tdist.Normal(
+            torch.tensor([0.0] * batch_size, device=device),
+            torch.tensor([1.0] * batch_size, device=device),
+        )
+
+        # Define a method conditionally for estimating instantaneous pump rate
+        if pump_rate_flag:
+            pump_field = lambda n: pump * (n + 1) / iterations
+        else:
+            pump_field = lambda n: pump
+
+        # Perform the solve over the specified number of iterations
+
+        for i in range(iterations):
+            c_drift = self.calculate_drift(c, pump_field(i), S)
+
+            wiener_increment_c = wiener_dist_c.sample((problem_size,)).transpose(
+                0, 1
+            ) * np.sqrt(dt)
+
+            c += dt * feedback_scale * c_drift + sigma * wiener_increment_c
+
+            # Ensure variables are within any problem constraints
+            # The lower bound is determined by ell=-S, and upper bound by u=S
+            c = self.fit_to_constraints(c, -S, S)
+
+            # If evolution_step_size is specified, save the values if this iteration
+            # aligns with the step size or if this is the last iteration
+            if evolution_step_size and (
+                i % evolution_step_size == 0 or i + 1 >= iterations
+            ):
+                # Update the record of the sample values with the values found at
+                # this iteration
+                self.c_sample[
+                    :, :, samples_taken
+                ] = c  # TODO: Consult if c needs calibration (c+S)/(2*S)
```

Comment:
Same here

Reply:
What do you think @FarhadK-1QBit about this?

Reply:
The calibration is not necessary here. `c` can be stored as is in the `self.c_sample` variable.